### ISSUE TITLE
Support PreBuildTasks

### ISSUE THREAD

**Issue by caendesilva** [hydephp/develop#1153]
The build tasks are kindof in a limbo, where they are named BuildTasks, but we also use the name PostBuildTasks. To be honest, we should probably support PreBuildTasks as well.

**Comment by caendesilva** [comment]
Maybe a good way to do this is a simple set of interfaces. RunsBeforeBuild and RunsAfterBuild.

**Comment by caendesilva** [comment]
> Maybe a good way to do this is a simple set of interfaces. RunsBeforeBuild and RunsAfterBuild.

With this, question is what should happen if a build task has no interface?

**Comment by caendesilva** [comment]
Better idea from Robert:

> Default methods do nothing, you overwrite the method in your class to do what needs to be done 
> ... you have a base class where both these methods do nothing ... then in your class, you implement 1 or the other (or both, if needed) to do something

**Comment by caendesilva** [comment]
Actually, I think the interface system is nicer, and feels more like Laravel and how they handle jobs. 

> With this, question is what should happen if a build task has no interface?
This is still an issue though

**Comment by caendesilva** [comment]
### Feedback from ChatGPT:

### Prompt

I had an idea that the task can implement one of two (empty) interfaces, either RunsBeforeBuild or RunsAfterBuild.

I kinda like this system, but it brings two problems: What if neither interface is implemented, and what if both interfaces are implemented? How should I handle that?

### Answer

Using interfaces to indicate whether a task should run before or after the build is a good idea. To handle the cases where a task implements neither or both interfaces, you could add some additional logic to the framework that checks for these cases and handles them appropriately.

For example, if a task implements neither interface, you could assume that it should run after the build. Alternatively, you could throw an exception or log a warning to alert the user that the task's behavior is ambiguous.

Here's an example of how this might look in code:

```php
interface RunsBeforeBuild {
    // This interface is empty
}

interface RunsAfterBuild {
    // This interface is empty
}

abstract class BuildTask {
    abstract public function run(): void;

    public function shouldRunBeforeBuild(): bool {
        return $this instanceof RunsBeforeBuild;
    }

    public function shouldRunAfterBuild(): bool {
        return $this instanceof RunsAfterBuild;
    }
}

// Usage:
$tasks = [
    new MyTask(), // Assumes MyTask should run after the build
    new MyPreBuildTask(), // Implements RunsBeforeBuild interface
    new MyPostBuildTask(), // Implements RunsAfterBuild interface
    new MyAmbiguousTask(), // Does not implement either interface
];

foreach ($tasks as $task) {
    if ($task->shouldRunBeforeBuild()) {
        // Run task before build
    } elseif ($task->shouldRunAfterBuild()) {
        // Run task after build
    } else {
        // Task behavior is ambiguous - handle error/warning as appropriate
    }
}
```

If a task implements both RunsBeforeBuild and RunsAfterBuild, you could either throw an exception or log a warning to alert the user that the task's behavior is ambiguous. Alternatively, you could provide some kind of configuration option or flag that allows the user to specify which behavior to use.

**Comment by caendesilva** [comment]
My idea was to handle separate arrays for pre/post types, that has the benefit of a clear data structure when dumping the class, but ChatGPTs way is better.

With Robert's idea I was unsure how to implement the instances, but that could be solved with the GPT loop idea.

**Comment by caendesilva** [comment]
> I was unsure how to implement the instances because we'd need to instantiate them twice

Resolved in 9318d0a8a8c23e32c9d8d73999a9d47cc6e1466c
